rllib/core/models/tf/encoder.py

```diff
@@ -135,13 +105,145 @@ def _forward(self, inputs: NestedDict, **kwargs) -> NestedDict:
+            }
+        )
+
+
+class TfLSTMEncoder(TfModel, Encoder):
+    """An encoder that uses an LSTM cell and a linear layer."""
+
+    def __init__(self, config: LSTMEncoderConfig) -> None:
         TfModel.__init__(self, config)
-        ActorCriticEncoder.__init__(self, config)
+
+        # Create the tf LSTM layers.
+        self.lstms = []
+        for _ in range(config.num_lstm_layers):
+            self.lstms.append(
+                tf.keras.layers.LSTM(
+                    config.hidden_dim,
+                    time_major=not config.batch_major,
+                    use_bias=config.use_bias,
+                    return_sequences=True,
+                    return_state=True,
+                )
+            )
+
+        # Create the final dense layer.
+        self.linear = tf.keras.layers.Dense(
+            units=config.output_dims[0],
+            use_bias=config.use_bias,
+        )
+
+    @override(Model)
+    def get_input_specs(self) -> Optional[Spec]:
+        return SpecDict(
+            {
+                # b, t for batch major; t, b for time major.
+                SampleBatch.OBS: TfTensorSpec("b, t, d", d=self.config.input_dims[0]),
+                STATE_IN: {
+                    "h": TfTensorSpec(
+                        "b, l, h",
+                        h=self.config.hidden_dim,
+                        l=self.config.num_lstm_layers,
+                    ),
+                    "c": TfTensorSpec(
+                        "b, l, h",
+                        h=self.config.hidden_dim,
+                        l=self.config.num_lstm_layers,
+                    ),
+                },
+            }
+        )
+
+    @override(Model)
+    def get_output_specs(self) -> Optional[Spec]:
+        return SpecDict(
+            {
+                ENCODER_OUT: TfTensorSpec("b, t, d", d=self.config.output_dims[0]),
+                STATE_OUT: {
+                    "h": TfTensorSpec(
+                        "b, l, h",
+                        h=self.config.hidden_dim,
+                        l=self.config.num_lstm_layers,
+                    ),
+                    "c": TfTensorSpec(
+                        "b, l, h",
+                        h=self.config.hidden_dim,
+                        l=self.config.num_lstm_layers,
+                    ),
+                },
+            }
+        )
+
+    @override(Model)
+    def get_initial_state(self):
+        return {
+            "h": tf.zeros((self.config.num_layers, self.config.hidden_dim)),
+            "c": tf.zeros((self.config.num_layers, self.config.hidden_dim)),
+        }
+
+    @override(Model)
+    def _forward(self, inputs: NestedDict, **kwargs) -> NestedDict:
+        out = tf.cast(inputs[SampleBatch.OBS], tf.float32)
+
+        # States are batch-first when coming in. Make them layers-first.
+        states_in = tree.map_structure(
+            lambda s: tf.transpose(s, perm=[1, 0] + list(range(2, len(s.shape)))),
+            inputs[STATE_IN],
+        )
```

Comment:
Why are we doing this instead of doing `perm=[1, 0, 2]`?
All `state in`s will always have a single dimension right? Otherwise specs validation would also vail.
I think this code is "too" general and can be simplified to fit this particular use case, right?

Reply:
done